### PR TITLE
[optim] Merge the pyi files into py files of optimizer

### PR DESCRIPTION
**Pull request by david20571015**
Continue the work of pytorch/pytorch#125153

### CODE REVIEW

File: torch/optim/adam.py [hunk 1]
```diff
@@ -708,16 +716,17 @@ def _fused_adam(
...
+        {found_inf.device: found_inf} if found_inf is not None else {}
     )
-    found_inf_dict = {found_inf.device: found_inf} if found_inf is not None else None
 
     # We only shuffle around the lr when it is a Tensor and on CUDA, otherwise, we prefer
     # treating it as a scalar.
-    lr_dict = (
-        {lr.device: lr} if isinstance(lr, Tensor) and str(lr.device) != "cpu" else None
-    )
+    lr = torch.tensor(lr)
```
Comment (janeyx99) [inline torch/optim/adam.py:728]: why is this necessary? we don’t want to wrap lr into a Tensor normally

File: torch/optim/adamw.py [hunk 2]
```diff
@@ -27,10 +29,10 @@ class AdamW(Optimizer):
     def __init__(
         self,
         params: ParamsT,
-        lr: Union[float, Tensor] = 1e-3,
+        lr=1e-3,
```
Comment (janeyx99) [inline torch/optim/adamw.py:32]: why the removal of types here?

File: torch/optim/asgd.py [hunk 3]
```diff
@@ -376,6 +377,7 @@ def _multi_tensor_asgd(
             torch._foreach_add_(grouped_state_steps, 1)
 
         # intermediate = grad + param * lambd
+        intermediate: Union[Tuple[Tensor, ...], List[Tensor]]
```
Comment (janeyx99) [inline torch/optim/asgd.py:380]: what’s the diff between defining this here vs the first time it’s instantiated?

**Review by janeyx99 — COMMENTED** [review]
Started reviewing, but it’d be easier to review if you could split the lr_scheduler changes from the optims

File: torch/optim/adam.py (hunk 1 above)
Comment (david20571015) [reply torch/optim/adam.py:728]: I believe my modification here may be incorrect, I will fix this.

File: torch/optim/adamw.py (hunk 2 above)
Comment (david20571015) [reply torch/optim/adamw.py:32]: reverted
I accidently removed it.

File: torch/optim/asgd.py (hunk 3 above)
Comment (david20571015) [reply torch/optim/asgd.py:380]: without this, mypy will raise an error `Incompatible types in assignment (expression has type "tuple[Tensor, ...]", variable has type "list[Tensor]")` in line 386, 392 and 411 because `intermediate` was assigned as list[Tensor] in line 384. I think it is more concise to assign a union type than add 3 type ignores.

**Comment by david20571015** [comment]
thanks for your review @janeyx99, I splited the changes about lr_scheduler into #125556.

File: torch/optim/adam.py [hunk 4]
```diff
@@ -27,10 +29,10 @@ class Adam(Optimizer):
     def __init__(
         self,
         params: ParamsT,
-        lr: Union[float, Tensor] = 1e-3,
+        lr=1e-3,
```
Comment (janeyx99) [inline torch/optim/adam.py:32]: looks like there are more places you accidentally removed

File: torch/optim/adamw.py [hunk 5]
```diff
@@ -701,16 +710,17 @@ def _fused_adamw(
...
+        {found_inf.device: found_inf} if found_inf is not None else {}
     )
-    found_inf_dict = {found_inf.device: found_inf} if found_inf is not None else None
 
     # We only shuffle around the lr when it is a Tensor and on CUDA, otherwise, we prefer
     # treating it as a scalar.
-    lr_dict = (
-        {lr.device: lr} if isinstance(lr, Tensor) and str(lr.device) != "cpu" else None
-    )
+    lr = torch.tensor(lr)
```
Comment (janeyx99) [inline torch/optim/adamw.py:722]: remove

File: torch/optim/adamw.py [hunk 6]
```diff
@@ -688,11 +697,11 @@ def _fused_adamw(
     amsgrad: bool,
     beta1: float,
     beta2: float,
-    lr: Union[float, Tensor],
+    lr: Union[Tensor, float],
     weight_decay: float,
     eps: float,
     maximize: bool,
-    capturable: bool,  # Needed for consistency.
```
Comment (janeyx99) [inline torch/optim/adamw.py:695]: why delete the comment?

File: torch/optim/swa_utils.py [hunk 7]
```diff
@@ -242,15 +238,8 @@ def update_parameters(self, model: Module):
...
-                grouped_tensors = _group_tensors_by_device_and_dtype(
-                    cast(TensorListList, [self_param_detached, model_param_detached])
-                )
-                grouped_tensors = cast(
-                    Dict[
-                        Tuple[torch.device, torch.dtype],
-                        Tuple[List[List[Tensor]], Indices],
-                    ],
-                    grouped_tensors,
+                grouped_tensors = Optimizer._group_tensors_by_device_and_dtype(
```
Comment (janeyx99) [inline torch/optim/swa_utils.py:241]: Why the switch to use the optimizer one here?

**Review by janeyx99 — COMMENTED** [review]
Almost there!

File: torch/optim/adam.py (hunk 4 above)
Comment (david20571015) [reply torch/optim/adam.py:32]: fixed. sorry for this.

File: torch/optim/adamw.py (hunk 5 above)
Comment (david20571015) [reply torch/optim/adamw.py]: fixed

File: torch/optim/adamw.py (hunk 6 above)
Comment (david20571015) [reply torch/optim/adamw.py:695]: I thought this is unnecessary because `has_complex` didn't have this comment. But I add it back and add a same comment for `has_complex`.

File: torch/optim/swa_utils.py (hunk 7 above)
Comment (david20571015) [reply torch/optim/swa_utils.py:241]: The optimizer one internally calls the original function and also supports compiling. It doesn't need to perform a lot of type casting because it is ignored in the optimizer one. Is this not suitable here?"

File: torch/optim/adam.py [hunk 8]
```diff
@@ -29,8 +31,8 @@ def __init__(
         params: ParamsT,
         lr: Union[float, Tensor] = 1e-3,
         betas: Tuple[float, float] = (0.9, 0.999),
-        eps: float = 1e-8,
-        weight_decay: float = 0,
+        eps=1e-8,
+        weight_decay=0,
```
Comment (janeyx99) [inline torch/optim/adam.py:35]: here too

File: torch/optim/adamw.py [hunk 9]
```diff
@@ -29,8 +31,8 @@ def __init__(
         params: ParamsT,
         lr: Union[float, Tensor] = 1e-3,
         betas: Tuple[float, float] = (0.9, 0.999),
-        eps: float = 1e-8,
-        weight_decay: float = 1e-2,
+        eps=1e-8,
+        weight_decay=1e-2,
```
Comment (janeyx99) [inline torch/optim/adamw.py:35]: here

File: torch/optim/swa_utils.py (hunk 7 above)
Comment (janeyx99) [reply torch/optim/swa_utils.py:241]: eh...the one in optimizer is a special case where it gets skipped in compile world..I'd rather not use it here if not necessary yet.

File: torch/optim/adam.py (hunk 8 above)
Comment (janeyx99) [reply torch/optim/adam.py:35]: though I see other places where this wasn't necessary--do you know the difference between defining vs not here?

If it's fine to not have a type, I am happy with not having one.

**Review by janeyx99 — APPROVED** [review]

File: torch/optim/swa_utils.py (hunk 7 above)
Comment (david20571015) [reply torch/optim/swa_utils.py:241]: Okay, I'll roll it back. Thanks for explaining this.

File: torch/optim/adam.py (hunk 8 above)
Comment (david20571015) [reply torch/optim/adam.py:35]: Sorry, could you please explain the difference to me? I think mypy interprets weight_decay as an int, so should I specify a float here or change 0 to 0.0?

Comment (janeyx99) [reply torch/optim/adam.py:35]: It does look like there's no error and it will allow an int or a float or anything (it might just infer it as Any). From my lil endeavor, it still feels the best to include the type. For example:

```
class Foo:
    def __init__(self, arg, kwarg=0):
        self.arg = arg
        self.kwarg = kwarg

a = Foo(1, "l")

print(5 + a.kwarg)
```
This will not error and will do the type conversion automatically.

But the intention is closer to: 
```
class Foo:
    def __init__(self, arg, kwarg: int = 0):
        self.arg = arg
        self.kwarg = kwarg

a = Foo(1, "l")

print(5 + a.kwarg)
```
And here, with the type, mypy will error properly at the instantiation of a

Comment (david20571015) [reply torch/optim/adam.py:36]: Okay, understood completely. I'll go ahead and add the types.

File: torch/optim/adamw.py [hunk 10]
```diff
@@ -318,92 +318,6 @@ def step(self, closure=None):
 )
 
 
-def adamw(
```
Comment (janeyx99) [inline torch/optim/adamw.py]: wait is this deletion intended?

Comment (david20571015) [reply torch/optim/adamw.py]: I merged main branch to this, and [0f02e0a](https://github.com/pytorch/pytorch/commit/0f02e0aa39b3dff15a8e2d1a81d049e838d7c696) moved the function to the end of the file.

**Comment by david20571015** [comment]
Sorry @janeyx99, I mistakenly thought the failures might been fixed, so I rebase to the main branch. However, after digging into the error message, I found that was caused by my code and I fixed it in https://github.com/pytorch/pytorch/pull/125452/commits/165fed5261170b85f70ae4e64365b7e77a0cea68. Could you approve the CI again please.

Sorry @janeyx99, I mistakenly thought the failures might have been fixed, so I rebased onto the main branch. However, after investigating the error message, I found that it was caused by my code, and I fixed it in https://github.com/pytorch/pytorch/pull/125452/commits/165fed5261170b85f70ae4e64365b7e77a0cea68. Could you please approve the CI again?

**Comment by david20571015** [comment]
Should I keep the parameter name `d_p_list`? None of the affected functions are exposed to external users.

image

**Comment by janeyx99** [comment]
> Should I keep the parameter name `d_p_list`? None of the affected functions are exposed to external users.

Yes, you should keep the name. sgd is publicly exposed and the argument is sadly not positional only.

**Comment by janeyx99** [comment]
The removal in optimizer `__init__.py` is the torch.optim.sgd module, not the function, which is offered so people can manage their own state. (I have yet to figure out why we do the dels there but that is not relevant for this discussion.)

**Comment by david20571015** [comment]
> > Should I keep the parameter name `d_p_list`? None of the affected functions are exposed to external users.
> 
> Yes, we should keep the name. sgd is publicly exposed and the argument is sadly not positional only.

done.

**Comment by janeyx99** [comment]
looks like there are merge conflicts now :/

**Comment by david20571015** [comment]
> looks like there are merge conflicts now :/

fixed

**Comment by janeyx99** [comment]
Please run lintrunner locally to ensure lint passes for the next commit. Getting lintrunner locally is pretty simple: just
```
pip install lintrunner
lintrunner init
lintrunner -a
```

**Comment by david20571015** [comment]
> Please run lintrunner locally to ensure lint passes for the next commit (and usually before pushing). Getting lintrunner locally is pretty simple: just
> 
> ```
> pip install lintrunner
> lintrunner init
> lintrunner -a
> ```

ok! fixed
